File: pegasusX/apps/driver-app-android/app/src/main/java/com/pegasusx/driver/data/telemetry/RouteNavigation.kt
package com.pegasusx.driver.data.telemetry

import com.pegasusx.driver.data.model.RouteStep
import com.pegasusx.driver.util.Haversine
import kotlin.math.roundToInt

const val NAVIGATION_STEP_PASSED_METERS = 35.0

data class NavigationCue(
    val instruction: String,
    val distanceM: Double,
    val maneuver: String? = null,
)

fun advanceNavigationStepIndex(
    currentIndex: Int,
    steps: List<RouteStep>,
    lat: Double,
    lng: Double,
): Int {
    if (steps.isEmpty()) {
        return 0
    }
    var lastPassedIndex = -1
    for (index in steps.indices) {
        val step = steps[index]
        val distance = Haversine.distanceMeters(lat, lng, step.lat, step.lng)
        if (distance <= NAVIGATION_STEP_PASSED_METERS) {
            lastPassedIndex = index
        }
    }
    if (lastPassedIndex < 0) {
        return currentIndex.coerceIn(0, steps.lastIndex)
    }
    return (lastPassedIndex + 1).coerceAtMost(steps.lastIndex)
}

fun resolveNavigationCue(
    steps: List<RouteStep>,
    stepIndex: Int,
    lat: Double,
    lng: Double,
): NavigationCue? {
    val step = steps.getOrNull(stepIndex) ?: return null
    val distance = Haversine.distanceMeters(lat, lng, step.lat, step.lng)
    return NavigationCue(
        instruction = step.instruction,
        distanceM = distance,
        maneuver = step.maneuver,
    )
}

fun formatNavigationDistance(meters: Double): String {
    return if (meters < 1000) {
        "${meters.roundToInt()} m"
    } else {
        String.format("%.1f km", meters / 1000.0)
    }
}

File: pegasusX/apps/driver-app-ios/driverappios/driverappios/Utilities/RouteNavigation.swift
//
//  RouteNavigation.swift
//  driverappios
//

import CoreLocation
import Foundation

struct NavigationCue: Equatable {
    let instruction: String
    let distanceM: Double
    let maneuver: String?
}

enum RouteNavigation {
    static let stepPassedMeters: CLLocationDistance = 35

    static func advanceStepIndex(
        currentIndex: Int,
        steps: [RouteStep],
        coordinate: CLLocationCoordinate2D
    ) -> Int {
        guard !steps.isEmpty else { return 0 }
        var lastPassedIndex = -1
        for index in steps.indices {
            let step = steps[index]
            let stepCoordinate = CLLocationCoordinate2D(latitude: step.lat, longitude: step.lng)
            let distance = haversineDistance(from: coordinate, to: stepCoordinate)
            if distance <= stepPassedMeters {
                lastPassedIndex = index
            }
        }
        if lastPassedIndex < 0 {
            return min(max(currentIndex, 0), steps.count - 1)
        }
        return min(lastPassedIndex + 1, steps.count - 1)
    }

    static func resolveCue(
        steps: [RouteStep],
        stepIndex: Int,
        coordinate: CLLocationCoordinate2D
    ) -> NavigationCue? {
        guard steps.indices.contains(stepIndex) else { return nil }
        let step = steps[stepIndex]
        let stepCoordinate = CLLocationCoordinate2D(latitude: step.lat, longitude: step.lng)
        let distance = haversineDistance(from: coordinate, to: stepCoordinate)
        return NavigationCue(
            instruction: step.instruction,
            distanceM: distance,
            maneuver: step.maneuver
        )
    }

    static func formatDistance(_ meters: Double) -> String {
        if meters < 1000 {
            return "\(Int(meters.rounded())) m"
        }
        return String(format: "%.1f km", meters / 1000.0)
    }
}
